include/api/CBoostedTreeRegressionInferenceModelBuilder.h

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+#ifndef INCLUDED_ml_api_CBoostedTreeRegressionInferenceModelBuilder_h
+#define INCLUDED_ml_api_CBoostedTreeRegressionInferenceModelBuilder_h
+
+#include <maths/CBoostedTree.h>
+
+#include <api/CInferenceModelDefinition.h>
+#include <api/ImportExport.h>
+
+#include <rapidjson/document.h>
+
+#include <string>
+
+namespace ml {
+namespace api {
+
+class API_EXPORT CBoostedTreeRegressionInferenceModelBuilder
+    : public maths::CBoostedTree::Visitor {
+public:
+    using TDoubleVec = std::vector<double>;
+    using TStringVec = std::vector<std::string>;
+    using TStringSizeUMap = std::unordered_map<std::string, std::size_t>;
+    using TStringSizeUMapVec = std::vector<TStringSizeUMap>;
+    using TSizeStringUMap = std::unordered_map<std::size_t, std::string>;
+    using TSizeStringUMapVec = std::vector<TSizeStringUMap>;
+
+public:
+    CBoostedTreeRegressionInferenceModelBuilder(const TStringVec& fieldNames,
+                                                const TStringSizeUMapVec& categoryNameMap);
+    void visit(const maths::CBoostedTree* tree) override;
+
+    void visit(const maths::CBoostedTreeImpl* impl) override;
+
+    void visit(const maths::CBoostedTreeNode* node) override;
```

Comment:
One more observation...

These visit methods are never called *anywhere*. Even if they were overridden to do something they wouldn't have any effect. This feels wrong, i.e. they should at least be called from accept methods. There seem like a couple of options:
1. Ditch the visit methods
2. Have a visit std::vector<TBoostedTreeNodeUPtr> for `addTree` and then provide accessors on `CBoostedTreeNode` to get all the relevant state call visit methods instead of `add*` methods and remove the later.

As it stands the way the model is visited is quite tied up to building the trained model representation and it makes the accept methods nice and clear to read. Given this, recasting everything in terms of `visit` seems a bit like premature generalisation, so I'd inclined to go with the first option.

Reply:
Ok. I remove the methods for the moment. I am not completely satisfied with the design, but as we get new use cases, we can generalize pragmatically.